packages/app/src/cli/services/dev/urls.ts

```diff
@@ -28,8 +28,17 @@ export interface ApplicationURLs {
   appProxy?: AppProxy
 }
 
-export interface FrontendURLOptions {
-  noTunnelUseLocalhost: boolean
+export type FrontendURLOptions = UseLocalhostFrontendUrlOptions | UseTunnelFrontendUrlOptions
+
+interface UseLocalhostFrontendUrlOptions {
+  noTunnelUseLocalhost: true
+  port: number
+  tunnelUrl?: undefined
+  tunnelClient?: undefined
+}
+
+interface UseTunnelFrontendUrlOptions {
+  noTunnelUseLocalhost: false
   tunnelUrl?: string
   tunnelClient: TunnelClient | undefined
 }
```

Comment:
I think this can be simplified with either 

```
interface UseLocalhostFrontendUrlOptions {
  noTunnelUseLocalhost: true
  port: number
}

interface UseTunnelFrontendUrlOptions {
  noTunnelUseLocalhost: false
  tunnelUrl?: string
  tunnelClient: TunnelClient | undefined
}
```

This simplified interfaces work in `generateFrontedURL` if you do an early return when `noTunnelUseLocalhost = true`

or just keep the old Interface and generate the port inside `generateFrontendURL`